docs/internals/optimizer.rst

```diff
@@ -1382,3 +1391,59 @@ into
     }
 
 The LiteralRematerialiser should be run before this step.
+
+Codegen-Based Optimizer Module
+==============================
+
+Currently, the codegen-based optimizer module consists of two distinct steps.
+
+The first step moves literals to the right of commutative binary operators which
+helps exploiting associativity.
+
+The second step uses unchecked arithmetic when generating code for incrementing
+the counter variable of specific ``for`` loops. This
+avoids wasting gas by identifying some conditions that guarantee
+that the counter variable cannot overflow. This eliminates the need to use an explicit
+unchecked arithmetic block inside the loop body to increment the counter variable.
+
+.. _unchecked-loop-optimizer:
+
+Unchecked Loop Increment
+------------------------
+
+Introduced in Solidity ``0.8.22``, this optimizer step is concerned with identifying
+the conditions under which the ``for`` loop counter can be safely incremented
+without overflow checks.
+
+This optimization is **only** applied to ``for`` loops with the general form:
+
+.. code-block:: solidity
+
+    for (uint i = X; i < Y; ++i) {
+        // variable i is not modified in the loop body
+    }
+
+The condition and the fact that the counter variable is only ever incremented
+guarantee that it never overflows.
+The precise requirements for the loop to be eligible for the optimization are as follows:
+
+- The loop condition is a comparison of the form ``i < Y``, for a local counter variable ``i`` and an expression ``Y``.
+- The built-in operator ``<`` is necessarily used in the loop condition. User-defined operators are **not** eligible.
+- The loop expression is a prefix or postfix increment of the counter variable, i.e, ``i++`` or ``++i``.
+- The loop counter is a local declared variable of a built-in integer type.
```

Comment:
```suggestion
- The loop counter is a locally declared variable of a built-in integer type.
```

Reply:
Small nitpick but "local declared" reads weird. Maybe "local variable" or "locally declared variable"?

Reply:
Yeah, it is weird, I guess I wanted to write locally declared. But I think "local variable" is better. Thanks!